crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -702,15 +779,39 @@ impl CumulocityConverter {
         match &channel {
             Channel::EntityMetadata => {
                 if let Ok(register_message) = EntityRegistrationMessage::try_from(message) {
-                    if let Err(e) = self.entity_store.update(register_message) {
+                    // Generate the c8y external id, if an external id is not provided in the message
+                    let external_id = register_message
+                        .entity_id
+                        .as_ref()
+                        .map(|v| v.to_string().into())
+                        .unwrap_or_else(|| self.map_to_c8y_external_id(&source));
```

Comment:
It's fragile to let this logic to the mapper. The only piece of code that is related to c8y is the `map_to_c8y_external_id`  method. See my comment on the `entity_store.update()` method.